### PR TITLE
Openshift graph refresh

### PR DESCRIPTION
**Pull request by cben**
- [x] Depended on https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/57 [merged]

- [x] Running all VCR tests, both older cassette and the newer deletion tests.

RFE: https://bugzilla.redhat.com/show_bug.cgi?id=1470021

### CODE REVIEW

**Comment by cben** [comment]
This is now ready.  @enoodle @zakiva @zeari @Ladas @agrare please review.

Got delayed as enabling the deletion tests on graph refresh revealed some issues, notably with `get_container_images=false` mode, and extending that test revealed some more...  (Yay tests!)

- One tiny point I'm not fixing in this PR, because it's also this way with old refresh, is images `tag` column — in `get_container_images=false` mode it's not preserved like the rest of the metadata but "regresses" to the kubernetes parser value.
  Fixing that in graph refresh would be easy matter of `blacklist_attributes!`.  But I really want exactly same behavior.  Fixing in both would be uglier: eca22601f461b7585aeafa0dcd2c7b413203d8f2.

P.S. `get_container_images=false` mode is getting annoying to support, I would like to _drop_ it in favor of `store_unused_images=false` mode proposed in #9.

P.S. @Ladas has some more fixes but they're not covered by current specs.  We'll do them in later PR(s).

**Comment by cben** [comment]
ping . @enoodle @zakiva @zeari please review.

**Comment by agrare** [comment]
> P.S. `get_container_images=false` mode is getting annoying to support, I would like to drop it in favor of `store_unused_images=false` mode proposed in #9.

I'm good with that now that we have a way of increasing the timeout on the collection

**Comment by cben** [comment]
@moolitayer @enoodle @yaacov @zgalor @zakiva @zeari Please review — this is blocking delivering graph refresh for Scale team to test, and some upcoming PRs...

File: spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb [hunk 1]
```diff
@@ -468,3 +491,29 @@ def assert_container_node_with_no_hawk_attributes
...
+  context "classical refresh" do
+    before(:each) do
+      stub_settings_merge(
+        :ems_refresh => {:openshift => {:inventory_object_refresh => false}}
+      )
+
+      expect(ManageIQ::Providers::Openshift::ContainerManager::RefreshParser).not_to receive(:ems_inv_to_inv_collections)
+    end
+
+    include_examples "openshift refresher VCR tests"
```
Comment (moolitayer) [inline spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb]: :+1:

File: app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb [hunk 2]
```diff
@@ -10,9 +10,34 @@ def ems_inv_to_hashes(inventory, options = Config::Options.new)
...
+      def initialize_inventory_collections(ems, options)
+        super
+        # get_container_images=false mode is a stopgap to speed up refresh by reducing functionality.  When it's flipped from true to false, we should at least retain existing metadata.
+        if options.get_container_images
+          initialize_custom_attributes_collections(ems.container_images, %w(labels docker_labels))
+        end
+      end
+
+      # We rely on Kubernetes initialize_inventory_collections setting up our collections too.
+      # TODO: does it mean kubernetes refresh would wipe out the openshift tables?
```
Comment (moolitayer) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: @cben can you explain why? initialize_inventory_collections is filling in some metadata (and the ems) why would that wipe tables?

Comment (Ladas) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: Yeah, if you would run kubernetes refresh on openshift provider, it would delete the tables you are not sending data for. But we don't do that (running another's manager type refresh), since that would break anywhere. :-)

File: app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb [hunk 3]
```diff
@@ -87,6 +112,98 @@ def get_templates(inventory)
...
 
+      ## InventoryObject refresh methods
+
+      def merge_projects_into_namespaces_graph(inventory)
+        collection = @inv_collections[:container_projects]
+
+        inventory["project"].each do |data|
+          h = parse_project(data)
+          # TODO: Assumes full refresh, and running after get_namespaces_graph.
+          #   Would be a problem with partial refresh.
```
Comment (moolitayer) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: Must run after get_namespaces_graph` seems like a fair assumption (we have lots of those)

Can you rephrase the TODO's as assertions? 
(I mean something like `# Must run after get_namespaces_graph`, `# need modification for partial refresh`, etc)
I'm not sure there is actually something todo here right now, so let's bring in as little `TODO` as possible.

File: app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb (hunk 2 above)
Comment (cben) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: The parser returns all `@inv_collections.values` as the collections to sync.
So it always includes e.g. `ems.container_templates`.  Persistor doesn't know it's empty because we didn't fill it, or empty because all templates were deleted in openshift, if it finds any `ems.container_templates` in DB it will delete them.

These collections should always be empty in DB for a kubernetes provider, so it's not really a problem.  I want to see where @Ladas's upcoming restructurings (including https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/73) end up before moving openshift collections here...

But already had to move image labels to `initialize_inventory_collections` here, for this very reason, so I'll move this comment to the `super` call there.

File: app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb (hunk 3 above)
Comment (cben) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: Old refresh had these assumptions (order & full refresh) everywhere.
Graph refresh is pretty close to not having any :-), and there are concrete wins to making them independent.
But yeah, there is more to track for partial refresh, I'll make assertions not TODOs.

File: app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb (hunk 2 above)
Comment (Ladas) [inline app/models/manageiq/providers/openshift/container_manager/refresh_parser.rb]: right, I moved the initialize_inventory_collections a bit, now it's under Openshift Persister 
https://github.com/ManageIQ/manageiq-providers-openshift/pull/39

**Comment by cben** [comment]
Rewrote comments, PTAL.

**Comment by cben** [comment]
Kicking Travis as master is now green

**Comment by moolitayer** [comment]
@cben travis failure seems to be related to ContainerDefinition removal

**Comment by cben** [comment]
Strange, I thought ContainerDefinition has been all fixed, checking...
